File: leetcode-cn/218.天际线问题.py
#
# @lc app=leetcode.cn id=218 lang=python3
#
# [218] 天际线问题
#

# @lc code=start
from typing import List


class Solution:
    def getSkyline(self, buildings: List[List[int]]) -> List[List[int]]:
        all_points = []
        for building in buildings:
            all_points.append((building[0], -building[2]))
            all_points.append((building[1], building[2]))

        all_points.sort()

        heights = []
        result = []
        heights.append(0)

        for x, h in all_points:
            if h < 0:
                h = -h
                max_height = max(heights)
                if h > max_height:
                    result.append([x, h])
                heights.append(h)
            else:
                before_max_height = max(heights)
                heights.remove(h)
                after_max_height = max(heights)
                if after_max_height != before_max_height:
                    result.append([x, after_max_height])

        return result


# @lc code=end

File: leetcode-cn/C++/218.天际线问题.cpp
/*
 * @lc app=leetcode.cn id=218 lang=cpp
 *
 * [218] 天际线问题
 */

// @lc code=start
#include <vector>
using namespace std;

class Solution
{
public:
    vector<vector<int>> getSkyline(vector<vector<int>> &buildings)
    {
        auto cmp = [](const pair<int, int> &a, const pair<int, int> &b) -> bool
        { return a.second < b.second; };
        priority_queue<pair<int, int>, vector<pair<int, int>>, decltype(cmp)> que(cmp);

        vector<int> boundaries;
        for (auto &building : buildings)
        {
            boundaries.emplace_back(building[0]);
            boundaries.emplace_back(building[1]);
        }
        sort(boundaries.begin(), boundaries.end());

        vector<vector<int>> ret;
        int n = buildings.size(), idx = 0;
        for (auto &boundary : boundaries)
        {
            while (idx < n && buildings[idx][0] <= boundary)
            {
                que.emplace(buildings[idx][1], buildings[idx][2]);
                idx++;
            }
            while (!que.empty() && que.top().first <= boundary)
            {
                que.pop();
            }

            int maxn = que.empty() ? 0 : que.top().second;
            if (ret.size() == 0 || maxn != ret.back()[1])
            {
                ret.push_back({boundary, maxn});
            }
        }
        return ret;
    }
};

// @lc code=end
